File: c/buildpy/src/project.c
#include "buildpy.h"

Project* project_new(void) {
    Project* project = malloc(sizeof(Project));
    if (!project) return NULL;

    project->shell = shell_new();
    if (!project->shell) {
        free(project);
        return NULL;
    }

    // Initialize directory paths
    char* cwd = shell_get_cwd(project->shell);
    if (!cwd) {
        shell_free(project->shell);
        free(project);
        return NULL;
    }

    snprintf(project->build_dir, sizeof(project->build_dir), "%s/build", cwd);
    snprintf(project->downloads_dir, sizeof(project->downloads_dir), "%s/build/downloads", cwd);
    snprintf(project->source_dir, sizeof(project->source_dir), "%s/build/src", cwd);
    snprintf(project->install_dir, sizeof(project->install_dir), "%s/build/install", cwd);
    snprintf(project->bin_dir, sizeof(project->bin_dir), "%s/build/install/bin", cwd);
    snprintf(project->lib_dir, sizeof(project->lib_dir), "%s/build/install/lib", cwd);
    snprintf(project->lib_static_dir, sizeof(project->lib_static_dir), "%s/build/install/lib", cwd);

    return project;
}

void project_free(Project* project) {
    if (project) {
        shell_free(project->shell);
        free(project);
    }
}

int project_setup(Project* project) {
    if (!project) return -1;

    log_info("Setting up build environment");

    const char* dirs[] = {
        project->build_dir,
        project->downloads_dir,
        project->source_dir,
        project->install_dir,
        project->bin_dir,
        project->lib_dir
    };

    for (size_t i = 0; i < sizeof(dirs) / sizeof(dirs[0]); i++) {
        if (!shell_exists(project->shell, dirs[i])) {
            log_info("Creating directory: %s", dirs[i]);
            if (shell_mkdir_p(project->shell, dirs[i]) != 0) {
                log_error("Failed to create directory: %s", dirs[i]);
                return -1;
            }
        }
    }

    log_info("Build environment setup complete");
    return 0;
}

void project_clean(Project* project) {
    if (!project) return;

    log_info("Cleaning build environment");

    if (shell_exists(project->shell, project->build_dir)) {
        log_info("Removing build directory: %s", project->build_dir);
        shell_rmtree(project->shell, project->build_dir);
    }

    log_info("Build environment cleaned");
}

File: ml/buildpy/lib/project.ml

type project = {
  cwd: string;
  build: string;
  downloads: string;
  src: string;
  install: string;
}

let create_project () =
  let cwd = Sys.getcwd () in
  let build = Filename.concat cwd "build" in
  let downloads = Filename.concat build "downloads" in
  let src = Filename.concat build "src" in
  let install = Filename.concat build "install" in
  { cwd; build; downloads; src; install }

let setup project =
  Shell.makedirs [project.build; project.downloads; project.src; project.install]

let clean project =
  if Sys.file_exists project.src then (
    Shell.log_info ("Cleaning src directory: " ^ project.src);
    let cmd = "rm" in
    let args = ["-rf"; project.src] in
    Shell.run_command cmd args
  )

let reset project =
  if Sys.file_exists project.build then (
    Shell.log_info ("Resetting build directory: " ^ project.build);
    let cmd = "rm" in
    let args = ["-rf"; project.build] in
    Shell.run_command cmd args
  )